x/ibc/03-connection/types/version.go

```diff
@@ -1,96 +1,171 @@
+// procedure.
+func PickVersion(counterpartyVersions []string) (string, error) {
+	versions := GetCompatibleVersions()
+
+	for _, ver := range versions {
+		// check if the source version is supported by the counterparty
+		if counterpartyVer, found := FindSupportedVersion(ver, counterpartyVersions); found {
+			sourceIdentifier, sourceFeatures, err := UnpackVersion(ver)
+			if err != nil {
+				return "", err
 			}
 
-			if _, ok := counterpartyVerLookup[compatibleVersions[i]]; ok {
-				supportedVersions = append(supportedVersions, compatibleVersions[i])
+			_, counterpartyFeatures, err := UnpackVersion(counterpartyVer)
+			if err != nil {
+				return "", err
 			}
+
+			featureSet := GetFeatureSetIntersection(sourceFeatures, counterpartyFeatures)
+
+			version := CreateVersionString(sourceIdentifier, featureSet)
+			return version, nil
 		}
+	}
+
+	return "", sdkerrors.Wrapf(
+		ErrVersionNegotiationFailed,
+		"failed to find a matching counterparty version (%s) from the supported version list (%s)", counterpartyVersions, versions,
+	)
+}
 
-	case n < m:
-		for i := m - 1; i >= n; i-- {
-			compatibleVerLookup[compatibleVersions[i]] = true
+// GetFeatureSetIntersection returns the intersections of source feature set
+// and the counterparty feature set. This is done by iterating over all the
+// features in the source version and seeing if they exist in the feature
+// set for the counterparty version.
+func GetFeatureSetIntersection(sourceFeatureSet, counterpartyFeatureSet []string) (featureSet []string) {
+	for _, feature := range sourceFeatureSet {
+		if contains(feature, counterpartyFeatureSet) {
+			featureSet = append(featureSet, feature)
 		}
+	}
 
-		for i := n - 1; i >= 0; i-- {
-			counterpartyVerLookup[counterpartyVersions[i]] = true
-			compatibleVerLookup[compatibleVersions[i]] = true
+	return featureSet
+}
 
-			// check if we've seen any of the versions
-			if _, ok := compatibleVerLookup[counterpartyVersions[i]]; ok {
-				supportedVersions = append(supportedVersions, counterpartyVersions[i])
-			}
+// VerifyProposedFeatureSet verifies that the entire feature set in the
+// proposed version is supported by this chain.
+func VerifyProposedFeatureSet(proposedVersion, supportedVersion string) bool {
+	_, proposedFeatureSet, err := UnpackVersion(proposedVersion)
+	if err != nil {
+		return false
+	}
 
-			if _, ok := counterpartyVerLookup[compatibleVersions[i]]; ok {
-				supportedVersions = append(supportedVersions, compatibleVersions[i])
-			}
+	_, supportedFeatureSet, err := UnpackVersion(supportedVersion)
+	if err != nil {
+		return false
+	}
+
+	for _, proposedFeature := range proposedFeatureSet {
+		if !contains(proposedFeature, supportedFeatureSet) {
+			return false
 		}
 	}
 
-	if len(supportedVersions) == 0 {
-		return ""
+	return true
+}
+
+// contains returns true if the provided string element exists within the
+// string set.
+func contains(elem string, set []string) bool {
```

Comment:
can we use `strings.Contains` instead?

Reply:
how do you mean? they take in different arguments. I cannot treat the features as one large string because then substring edge cases could occur ie is "feature0" in "feature0withZK" yes, but really I wanted just "feature0". Does that make sense?